### PR TITLE
Refactor retry logic for TransportMasterNodeAction

### PR DESCRIPTION
**Pull request by ywelsch**
Changes:
- Introduce inner class `AsyncSingleAction` to store state used across retry invocations.
- Same as for `TransportInstanceSingleOperationAction` and `TransportReplicationAction`, `onClusterServiceClose` consistently throws a `NodeClosedException` now.
- Added retry logic if master could not publish cluster state or stepped down before publishing (ZenDiscovery). The test `IndexingMasterFailoverIT` shows the issue (but will not be part of the final commit).
- Simplified retry logic by moving bits from different places into shared retry method.
  - Removed boolean flag `retrying` that aborted retrying after a single master node change (now we retry until timeout).
  - Two existing predicates that deal with master node changes unified in a single predicate `masterNodeChangedPredicate`.

Other implementation notes: In contrast to `TransportInstanceSingleOperationAction` and `TransportReplicationAction`, `onTimeout` does not try to execute the action a last time as in this case we wait for specific cluster change events. This keeps the behaviour consistent with the old code that directly failed on timeout.

### CODE REVIEW

**Comment by ywelsch** [comment]
The method `processBeforeDelegationToMaster` can also be removed. AFAICS, it is nowhere used.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 1]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
+                        }
+                    });
                 }
-                return;
             }
-            processBeforeDelegationToMaster(request, clusterState);
-            transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
+        }
+
+        private ClusterStateObserver.ChangePredicate retryableOrNoBlockPredicate() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: can we make these final fields on the parent class? they don't have any state.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 2]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
+                    ActionListener<Response> delegate = new ActionListener<Response>() {
+                        @Override
+                        public void onResponse(Response response) {
+                            listener.onResponse(response);
+                        }
+
+                        @Override
+                        public void onFailure(Throwable t) {
+                            if (t instanceof Discovery.FailedToCommitClusterStateException
+                                    || (t instanceof EsRejectedExecutionException && t.getMessage().startsWith("no longer master."))) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: lets' introduce a dedicated exception for this. We can upgrade discovery.zen.NotMasterException to be in the cluster package and use that.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 3]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
+                                logger.debug("master could not publish cluster state or stepped down before publishing, scheduling a retry");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: can we add the action to the log so we know which operation this is about?

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 4]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
-                                    return newState.nodes().masterNodeId() != null;
-                                }
-
-                                @Override
-                                public boolean apply(ClusterChangedEvent event) {
-                                    return event.nodesDelta().masterNodeChanged();
-                                }
+                    retry(new MasterNotDiscoveredException(), masterNodeChangedPredicate());
+                } else {
+                    processBeforeDelegationToMaster(request, clusterState);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: do we still need this method?

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 5]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
+                        public String executor() {
+                            return ThreadPool.Names.SAME;
+                        }
+
+                        @Override
+                        public void handleException(final TransportException exp) {
+                            Throwable cause = exp.unwrapCause();
+                            if (cause instanceof ConnectTransportException) {
+                                // we want to retry here a bit to see if a new master is elected
+                                logger.debug("connection exception while trying to forward request to master node [{}], scheduling a retry. Error: [{}]",
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: can we add the action name here too?

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 6]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
-                                }
-
-                                @Override
-                                public boolean apply(ClusterChangedEvent event) {
-                                    return event.nodesDelta().masterNodeChanged();
-                                }
+                    retry(new MasterNotDiscoveredException(), masterNodeChangedPredicate());
+                } else {
+                    processBeforeDelegationToMaster(request, clusterState);
+                    transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: using ActionListenerResponseHandler will simplify this lightly.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 7]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
+        }
 
+        private ClusterStateObserver.ChangePredicate masterNodeChangedPredicate() {
+            return new ClusterStateObserver.ChangePredicate() {
                 @Override
-                public void handleResponse(Response response) {
-                    listener.onResponse(response);
+                public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                     ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                    return newState.nodes().masterNodeId() != null && (previousState != newState || previousStatus != newStatus);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: I know this is how it used to be, but can we make the if be more like the `masterNodeChangePredicate` name and check the the master node is not null and have changed? (we now test for a cluster state change)

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 8]
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
-            });
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        listener.onFailure(new NodeClosedException(clusterService.localNode()));
+                    }
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        listener.onFailure(failure);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: can we add a timeout log here?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+public class IndexingMasterFailoverIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        final HashSet<Class<? extends Plugin>> classes = new HashSet<>(super.nodePlugins());
+        classes.add(MockTransportService.TestPlugin.class);
+        return classes;
+    }
+
+    public void testMasterFailoverDuringIndexingWithMappingChanges() throws Throwable {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: why remove this test?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        logger.info("--> start 4 nodes, 3 master, 1 data");
+        Settings sharedSettings = settingsBuilder()
+                .put(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES, 2)
+                .put(FaultDetection.SETTING_PING_TIMEOUT, "100ms") // speed things up
+                .put(ZenDiscovery.SETTING_PING_TIMEOUT, "200ms")
+                .put(DiscoverySettings.COMMIT_TIMEOUT, "100ms") // speed things up
+                .build();
+
+        for (int i = 0; i < 3; i++) {
+            internalCluster().startMasterOnlyNode(sharedSettings);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: startNodesAsync(3,shardSettings)?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        }
+
+        String dataNode = internalCluster().startDataOnlyNode(sharedSettings);
+
+        logger.info("--> wait for all nodes to join the cluster");
+
+        Client client = client();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: why cache the client?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+        assertThat(client.admin().cluster().prepareHealth().setWaitForNodes("4").get().isTimedOut(), equalTo(false));
+
+        // We index data with mapping changes into cluster and have master failover at same time
+        client(dataNode).prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
+
+        final BlockingQueue<Throwable> exceptions = new LinkedBlockingQueue();
+        Thread[] threads = new Thread[6];
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: can't we make do with one background thread?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+                }
+            });
+
+            threads[j].setName("t_" + j);
+            threads[j].start();
+        }
+
+        for (int i = 0; i < 10; i++) {
+            // periodically interrupt communication between master and other nodes in cluster
+            Thread.sleep(1000);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: boo hoe! Let's start a disruption, wait for a new master via the other nodes, stop the disruption and repeat. Why the sleep?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+                    } catch (Throwable t) {
+                        exceptions.offer(t);
+                    }
+                }
+            });
+
+            threads[j].setName("t_" + j);
+            threads[j].start();
+        }
+
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: I think we need a barrier.await() here too to make sure the threads and disruptions start together?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                    try {
+                        barrier.await();
+
+                        for (int i = 0; i < 50; i++) {
+                            // index data with mapping changes
+                            IndexResponse response = client(dataNode).prepareIndex("myindex", "mytype").setSource("field_" + threadId + "_" + i, "val").get();
+                            assertThat(response.isCreated(), equalTo(true));
+                        }
+                    } catch (Throwable t) {
+                        exceptions.offer(t);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: we can throw exceptions on the background thread and the test should fail with uncaught exceptions on a thread.

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            internalCluster().setDisruptionScheme(partition);
+
+            partition.startDisrupting();
+            Thread.sleep(1000);
+            partition.stopDisrupting();
+        }
+
+        for (Thread t : threads) t.join();
+
+        if (!exceptions.isEmpty()) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: can we also verify all the documents are found in the index?

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java [hunk 17]
```diff
@@ -320,4 +322,35 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
...
+
+    public void testMasterFailoverAfterStepDown() throws ExecutionException, InterruptedException {
+        Request request = new Request();
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+
+        final Response response = new Response();
+
+        clusterService.setState(ClusterStateCreationUtils.state(localNode, localNode, allNodes));
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: can we increase the timeout on the request? if one runs a debugger the test may fail to retry because a timeout happens, making it confusing.

**Comment by bleskes** [comment]
This looks great. Left a bunch of comments.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: retryableOrNoBlockPredicate depends on the current request being made. I chose a method so that the predicate object does not have to be created for every MasterNodeAction.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: I see. I don't think it's that bad to create it for every (master) request and keep it simple but up to you . If you leave like this please add a comment in the code as to why. Also, my comment referred to both this and masterNodeChangedPredicate , so fi you decided to change, please do it there too.

**Comment by ywelsch** [comment]
@bleskes, can you have another look?

I made (all) changes according to the provided suggestions, in particular:

- removed processBeforeDelegationToMaster
- changed methods retryableOrNoBlockPredicate() and masterNodeChangedPredicate() to fields
- changed masterNodeChangedPredicate implementation to only trigger on master node id changes
- moved discovery.zen.NotMasterException to cluster.NotMasterException
- log output in case of timeout while retrying (log-level warn)
- simplified IntegrationTest (took settings from DiscoveryWithServiceDisruptionsIT)

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 18]
```diff
@@ -48,6 +50,19 @@
  * A base class for operations that needs to be performed on the master node.
  */
 public abstract class TransportMasterNodeAction<Request extends MasterNodeRequest, Response extends ActionResponse> extends HandledTransportAction<Request, Response> {
+    private static final ClusterStateObserver.ChangePredicate masterNodeChangedPredicate = new ClusterStateObserver.ChangePredicate() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: nit: can we move this next to the other predicate? I think it's good to have them together.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 19]
```diff
@@ -75,152 +90,125 @@ protected boolean localExecute(Request request) {
...
-                                                       public boolean apply(ClusterChangedEvent event) {
-                                                           return event.nodesDelta().masterNodeChanged();
-                                                       }
-                                                   }
-                        );
-                    } else {
-                        listener.onFailure(exp);
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        logger.warn("timed out while retrying. Reason for retry: ", failure);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: Somewhat simpler: ("timed out while retrying [{}] after failure (timeout [{}])", action, failure) . I'm doubting between DEBUG and WARN for this log...

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+@ESIntegTestCase.SuppressLocalMode
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: I don't think we need this? we want to make sure we use zen, but that's it no?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+    public void testMasterFailoverDuringIndexingWithMappingChanges() throws Throwable {
+        logger.info("--> start 4 nodes, 3 master, 1 data");
+
+        final Settings sharedSettings = Settings.builder()
+                .put(FaultDetection.SETTING_PING_TIMEOUT, "1s") // for hitting simulated network failures quickly
+                .put(FaultDetection.SETTING_PING_RETRIES, "1") // for hitting simulated network failures quickly
+                .put("discovery.zen.join_timeout", "10s")  // still long to induce failures but to long so test won't time out
+                .put(DiscoverySettings.PUBLISH_TIMEOUT, "1s") // <-- for hitting simulated network failures quickly
+                .put("http.enabled", false) // just to make test quicker
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: I think we can clean up the http/transport/gatway settings here

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+        internalCluster().startMasterOnlyNodesAsync(3, sharedSettings).get();
+
+        String dataNode = internalCluster().startDataOnlyNode(sharedSettings);
+
+        logger.info("--> wait for all nodes to join the cluster");
+        ensureStableCluster(4);
+
+        // We index data with mapping changes into cluster and have master failover at same time
+        client(dataNode).prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: I think we can simplify this and make sure we have 1 shard, no replicas.

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+
+        final CyclicBarrier barrier = new CyclicBarrier(2);
+
+        Thread indexingThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    barrier.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: boo! :) just ignore :)

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+                    return;
+                } catch (BrokenBarrierException e) {
+                    e.printStackTrace();
+                    return;
+                }
+                for (int i = 0; i < 10; i++) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: is 10 iteration enough to reproduce reliably without your fix?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+                    // index data with mapping changes
+                    IndexResponse response = client(dataNode).prepareIndex("myindex", "mytype").setSource("field_" + i, "val").get();
+                    assertThat(response.isCreated(), equalTo(true));
+                }
+            }
+        });
+        indexingThread.setName("indexingThread");
+        indexingThread.start();
+
+        ensureStableCluster(4);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: why do this again?

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java (hunk 17 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: ping?

**Comment by bleskes** [comment]
Update looks great. I left some minor comments mostly around testing...

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java (hunk 18 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: yes, but then it cannot be static anymore (as inner class does not allow static member)

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: ok. fair enough..

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java (hunk 19 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: I was also torn between DEBUG and WARN. If you prefer, I can make it DEBUG for now, and we can switch it to WARN if that is useful for operations.

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 20 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: If I remove SuppressLocalMode, and add ("discovery.type", "zen") to the test, I hit the issue with zendiscovery + local mode again:

```
[2015-11-04 14:38:31,931][WARN ][org.elasticsearch.discovery] [node_t0] waited for 30s and no initial state was set by the discovery
[2015-11-04 14:38:31,932][INFO ][org.elasticsearch.node   ] [node_t0] started
[2015-11-04 14:38:32,042][WARN ][org.elasticsearch.discovery] [node_t1] waited for 30s and no initial state was set by the discovery
[2015-11-04 14:38:32,043][INFO ][org.elasticsearch.node   ] [node_t1] started
[2015-11-04 14:38:32,146][WARN ][org.elasticsearch.discovery] [node_t2] waited for 30s and no initial state was set by the discovery
```

and the following exception at the end:

```
MasterNotDiscoveredException[]
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:167)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.start(TransportMasterNodeAction.java:123)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:81)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:52)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:70)
	at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:47)
	at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:43)
	at org.elasticsearch.transport.local.LocalTransport.handleRequest(LocalTransport.java:290)
	at org.elasticsearch.transport.local.LocalTransport.messageReceived(LocalTransport.java:251)
	at org.elasticsearch.transport.local.LocalTransport$1.run(LocalTransport.java:229)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I am still investigating that one...

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 24 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: It triggered FailedToCommitClusterStateException every time on my machine.

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 25 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: extra stability (lol no, a relic from an earlier iteration)

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java (hunk 17 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: Maybe I misunderstood something here, but isn't that done by the first line in the method?
```
Request request = new Request().masterNodeTimeout(TimeValue.timeValueHours(1));
```

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java (hunk 18 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: fair enough, leave it or fair enough, change it ;-)

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 21 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: If I remove 

```
.put("transport.host", "127.0.0.1") // only bind on one IF we use v4 here by default
```

I run into #14513

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 20 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: argh, I forgot UnicastZenPing doesn't support local addresses by default. Sorry for the noise..  (we should fix this at one point)

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 24 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: ++

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 21 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: sure. I hope #14513 will be fixed soon. Not sure if you that is what you meant, but I think we better wait rather then work around it.

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java (hunk 23 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java]: if you want to know i happened, let's use logging..

File: core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java (hunk 17 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java]: oops. Missed it. You're right.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java (hunk 18 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: fair enough, leave it.

**Comment by ywelsch** [comment]
@bleskes Pushed another round addressing your comments. Rebased on master so I get #14535. I still cannot remove the line `.put("transport.host", "127.0.0.1")` as there seems to be another issue here. Ports are now correct but binding to multiple interfaces seems to mess with the test.

File: core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java [hunk 26]
```diff
@@ -48,6 +50,19 @@
  * A base class for operations that needs to be performed on the master node.
  */
 public abstract class TransportMasterNodeAction<Request extends MasterNodeRequest, Response extends ActionResponse> extends HandledTransportAction<Request, Response> {
+    private static final ClusterStateObserver.ChangePredicate masterNodeChangedPredicate = new ClusterStateObserver.ChangePredicate() {
+        @Override
+        public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                             ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+            return newState.nodes().masterNodeId() != null
+                    && !newState.nodes().masterNodeId().equals(previousState.nodes().masterNodeId());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java]: can we change this to support the case where master election started and completed between the sampling of the cluster state in the observer and when this function gets called? I think `newState.nodes().masterNodeId() != null
                    && newState != previousState);` is good? (plus a comment)

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+public class IndexingMasterDisruptionIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        final HashSet<Class<? extends Plugin>> classes = new HashSet<>(super.nodePlugins());
+        classes.add(MockTransportService.TestPlugin.class);
+        return classes;
+    }
+
+    public void testMasterDisruptionDuringIndexingWithMappingChanges() throws Throwable {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java]: can we add a short java doc about what the test does?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+            }
+        });
+        indexingThread.setName("indexingThread");
+        indexingThread.start();
+
+        barrier.await();
+
+        // interrupt communication between master and other nodes in cluster
+        // wait a little bit until some data mapping changes are queued
+        Thread.sleep(randomInt(500));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java]: why is this needed?

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+        otherNodes.remove(master);
+
+        NetworkPartition partition = new NetworkDisconnectPartition(Collections.singleton(master), otherNodes, random());
+        internalCluster().setDisruptionScheme(partition);
+
+        logger.info("--> disrupting network");
+        partition.startDisrupting();
+
+        logger.info("--> waiting for 3 seconds");
+        Thread.sleep(3000);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java]: why is this needed? ensureStableCluster(3,dataNode); should be enough imho...

File: core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+
+        logger.info("--> wait for all nodes to join the cluster");
+        ensureStableCluster(4);
+
+        // We index data with mapping changes into cluster and have master failover at same time
+        client().admin().indices().prepareCreate("myindex")
+                .setSettings(Settings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 0))
+                .get();
+        ensureGreen("myindex");
+        client().prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java]: why the extra index? can be removed no?

**Comment by bleskes** [comment]
@ywelsch did another round. Let's the leave the networking related code (i.e., `.put("transport.host", "127.0.0.1")` in and change them in another followup. As you indicated other tests have similar code so we can follow up and remove it from them as well.

**Comment by bleskes** [comment]
LGTM! :dancers:

**Comment by bleskes** [comment]
@ywelsch I think we should back port this to 2.2 .

**Comment by ywelsch** [comment]
2.2 does not have `FailedToCommitClusterStateException`...

**Comment by bleskes** [comment]
True, but it does have the “failed while in queue” logic. And the code is much cleaner.

**Comment by ywelsch** [comment]
Back ported and pushed to 2.x minus FailedToCommitClusterStateException.

**Comment by bleskes** [comment]
Thanks Yannick!
